### PR TITLE
Fix construction of HG frames to be 3D

### LINKED ISSUE #3350: Creating a HeliographicStonyhurst frame with SkyCoord does not always return a 3D frame
### Description
```python
from astropy.coordinates import SkyCoord
import astropy.units as u
from sunpy.coordinates import frames

SkyCoord([1,2,3,4], [1,3,4,5], unit=u.deg, frame=frames.HeliographicStonyhurst(obstime="2011-06-07T06:32:11.000"))
```
### Expected behaviour
The returned frame should be 3D with the rsun value as the distance coordinate.

### Actual behaviour

The frame is has a `UnitSpherical` representation:
```
<SkyCoord (HeliographicStonyhurst: obstime=2011-06-07T06:32:11.000): (lon, lat) in deg
    [(1., 1.), (2., 3.), (3., 4.), (4., 5.)]>
```

### PR DESCRIPTION
**Pull request by Cadair**
Fixes #3350

### CODE REVIEW

**Comment by ayshih** [comment]
While this PR does fix the issue, I suspect that there might be a better/cleaner way to fix it.  Or I could just be imagining things.  I'd like to do some tinkering before letting this through, unless a bugfix release is just around the corner.

**Comment by Cadair** [comment]
@ayshih It's not the end of the world if this floats for a little while, but it is breaking me messing around with #3353. 

I thought a decent amount about this, and I am pretty sure this is the best way. When you construct the frame with an explict `SphericalRepresentation` it seems to me the only logical thing to do is to throw it through our code to add R.

**Comment by ayshih** [comment]
I just PR-ed to your branch.  After poking around a bunch, I've decided that the whole approach to keyword checking and `represent_as` was unnecessarily complicated.  (The `represent_as` stuff is [my fault](https://github.com/sunpy/sunpy/commit/ec89603db6065af12784d1e31e9e41967e5f256b).)  Instead it will now always upgrade to 3D, as already promised in the docstring.  If a user chooses to specify a `representation_type` of `UnitSphericalRepresentation`, that's what they'll be shown, with component attributes to match, but the underlying data will be 3D.

**Comment by Cadair** [comment]
Do the changes to the `represent_as` function have any effect here? Is this still just a bug fix?

**Comment by ayshih** [comment]
In terms of user-facing API, the change here is what `.data` returns.  Previously, if a user specifically defined the `representation_type`, the class would try to preserve the underlying data and its representation, even if it was 2D rather than 3D.  The `.represent_as()` stuff was there so that the class was still able to work with the underlying data as if it were 3D.  Thus, previously, `.data` could still return a 2D representation, although this behavior was entirely undocumented.  (The bug identified in the corresponding issue was because the SkyCoord machinery can specify a `representation_type` that is different from the type of the representation.)  I think this still qualifies as a bug fix because these shenanigans were in violation of the docstring, which says that the data will always be converted to 3D.

**Comment by Cadair** [comment]
Could you perhaps suggest some appropriate wording for the changelog just to make that obvious to everyone? (or I can and then you can comment on it?)

**Comment by ayshih** [comment]
Here's a wordier version:
```
Fixed situations where 2D coordinates provided to `~sunpy.coordinates.frames.HeliographicStonyhurst` and `~sunpy.coordinates.frames.HeliographicCarrington` were not converted to 3D as intended.  Furthermore, the stored data will always be the post-conversion, 3D version.
```

File: changelog/3351.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed bug where Heliographic frames were not constructed as 3D as intended.
```
Comment (Cadair) [inline changelog/3351.bugfix.rst]:
```suggestion
Fixed situations where 2D coordinates provided to `~sunpy.coordinates.frames.HeliographicStonyhurst` and `~sunpy.coordinates.frames.HeliographicCarrington` were not converted to 3D as intended.  Furthermore, the stored data will always be the post-conversion, 3D version.
```

**Comment by Cadair** [comment]
Point of order, it seems that this backport failed because of the merge commit.

**Pull request merged**